src/bidiMapper/domains/script/scriptEvaluator.ts

```diff
@@ -251,154 +397,129 @@ export class ScriptEvaluator {
+                    resolver = resolve;
+                  });
+                  promise.then(() => queue.shift());
+                  return promise;
+                },
+
+                /**
+                 * Adds a message to the queue.
+                 * If the queue is initially empty, resolve the pending promise.
+                 */
+                sendMessage(message: string) {
+                  queue.push(message);
+                  if (queue.length === 1) {
+                    resolver(message);
+                  }
+                },
+              };
+            }),
+            returnByValue: false,
+            executionContextId: realm.executionContextId,
+            generateWebDriverValue: false,
+          }
         );
-    }
-  }
-
-  /**
-   * Serializes a given CDP object into BiDi, keeping references in the
-   * target's `globalThis`.
-   * @param cdpRemoteObject CDP remote object to be serialized.
-   * @param resultOwnership Indicates desired ResultOwnership.
-   * @param realm
-   */
-  async serializeCdpObject(
-    cdpRemoteObject: Protocol.Runtime.RemoteObject,
-    resultOwnership: Script.ResultOwnership,
-    realm: Realm
-  ): Promise<CommonDataTypes.RemoteValue> {
-    const arg = cdpRemoteObjectToCallArgument(cdpRemoteObject);
-
-    const cdpWebDriverValue: Protocol.Runtime.CallFunctionOnResponse =
-      await realm.cdpClient.sendCommand('Runtime.callFunctionOn', {
-        functionDeclaration: String((obj: unknown) => obj),
-        awaitPromise: false,
-        arguments: [arg],
-        generateWebDriverValue: true,
-        executionContextId: realm.executionContextId,
-      });
-    return realm.cdpToBidiValue(cdpWebDriverValue, resultOwnership);
-  }
+        const channelHandle = argEvalResult.result.objectId;
+
+        // Long-poll the message queue asynchronously.
+        (async () => {
+          for (;;) {
+            const message = await realm.cdpClient.sendCommand(
+              'Runtime.callFunctionOn',
+              {
+                functionDeclaration: String(async (channelHandle: any) => {
+                  return channelHandle.getMessage();
+                }),
+                arguments: [
+                  {
+                    value: channelHandle,
+                  },
+                ],
+                awaitPromise: true,
+                generateWebDriverValue: true,
+              }
+            );
+
+            this.#eventManager.registerEvent(
+              {
+                method: Script.EventNames.MessageEvent,
+                params: {
+                  channel: {
+                    type: 'channel',
+                    value: channelName,
+                  },
+                  data: await this.serializeCdpObject(
+                    {
+                      type: 'string',
+                      value: message.result,
+                    },
+                    'root',
+                    realm
+                  ),
```

Comment:
```suggestion
                  data: await realm.cdpToBidiValue(message, 'root'),
```

      result: await realm.cdpToBidiValue(
        cdpCallFunctionResult,
        resultOwnership
      ),